clippy_lints/src/swap.rs

```diff
@@ -174,53 +174,145 @@ fn check_manual_swap(cx: &LateContext<'_>, block: &Block<'_>) {
+        if let Some((lhs0, rhs0)) = parse(first)
+            && let Some((lhs1, rhs1)) = parse(second)
+            && is_same(cx, &lhs0, rhs1)
+            && is_same(cx, &lhs1, rhs0)
+        {
+            let lhs_opt: Option<Sugg<'_>> = if let ExprOrIdent::Expr(lhs_expr) = lhs0 {
+                Sugg::hir_opt(cx, lhs_expr)
+            } else {
+                None
+            };
+            let rhs_opt = Sugg::hir_opt(cx, rhs0);
+
+            let ident_l: Option<Ident> = if let ExprOrIdent::Ident(lhs_ident) = lhs0 {
+                Some(lhs_ident)
+            } else {
+                None
+            };
+
+            let (what, lhs, rhs) = if let (Some(first), Some(second)) = (lhs_opt, rhs_opt.clone()) {
+                (
+                    format!(" `{first}` and `{second}`"),
+                    first.mut_addr().to_string(),
+                    second.mut_addr().to_string(),
+                )
+            } else if let (Some(first_ident), Some(second)) = (ident_l, rhs_opt.clone()) {
+                (
+                    format!(" `{first_ident}` and `{second}`"),
+                    format!("&mut {first_ident}"),
+                    second.mut_addr().to_string(),
+                )
+            } else {
+                (String::new(), String::new(), String::new())
+            };
+
+            let span = if let StmtKind::Local(local_l) = first.kind && let StmtKind::Local(local_r) = second.kind {
+                local_l.span.to(local_r.span)
+            } else if let StmtKind::Local(local_l) = first.kind && let StmtKind::Semi(semi_r) = second.kind {
+                local_l.span.to(semi_r.span)
+            } else {
+                first.span.to(second.span)
+            };
+
+            let Some(sugg) = std_or_core(cx) else { return };
+            if !ExprOrIdent::span(first.span, second.span) { return };
+
+            if with_local(first, second) {
+                lint_almost_swapped_note(cx, span, &what, sugg, &lhs, &rhs);
+            } else {
+                lint_almost_swapped_suggest(cx, span, &what, sugg, &lhs, &rhs);
+            }
+        }
+    }
+}
+
+fn with_local<'a, 'hir>(stmt_l: &'a Stmt<'hir>, stmt_r: &'a Stmt<'hir>) -> bool {
+    if let StmtKind::Local(_) = stmt_l.kind && let StmtKind::Local(_) = stmt_r.kind {
+        true
+    } else if let StmtKind::Local(_) = stmt_l.kind && let StmtKind::Semi(_) = stmt_r.kind {
+        true
+    } else {
+        false
+    }
+}
 
-                let span = first.span.to(second.span);
-                let Some(sugg) = std_or_core(cx) else { return };
+fn is_same<'a, 'hir>(cx: &LateContext<'_>, lhs: &'a ExprOrIdent<'hir>, rhs: &'a Expr<'hir>) -> bool {
+    match lhs {
+        ExprOrIdent::Expr(expr) => eq_expr_value(cx, expr, rhs),
+        ExprOrIdent::Ident(ident) =>
+        if let ExprKind::Path(QPath::Resolved(None, path)) = rhs.kind &&
+            &path.segments.iter().map(|segment| segment.ident).collect::<Vec<_>>()[0] == ident {
```

Comment:
This isn't quite equivalent, it would say the path `foo::bar` is equal to `foo` as it's only checking the first segment, it should be something along the lines of `let [segment] = &path.segments && segment.ident == ident`